contracts/src/finance/vesting_wallet.rs

```diff
@@ -0,0 +1,621 @@
+    ///
+    /// # Panics
+    ///
+    /// If `total_allocation * (timestamp - self.start())` exceeds `U256::MAX`.
+    #[selector(name = "releasable")]
+    fn releasable_erc20(&mut self, token: Address)
+        -> Result<U256, Self::Error>;
+
+    /// Release the native tokens (Ether) that have already vested.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    ///
+    /// # Errors
+    ///
+    /// If Ether transfer fails, then the error [`Error::ReleaseEtherFailed`] is
+    /// returned.
+    ///
+    /// # Events
+    ///
+    /// Emits an [`EtherReleased`] event.
+    ///
+    /// # Panics
+    ///
+    /// If total allocation exceeds `U256::MAX`.
+    /// If `total_allocation * (timestamp - self.start())` exceeds `U256::MAX`.
+    #[selector(name = "release")]
+    fn release_eth(&mut self) -> Result<(), Self::Error>;
+
+    /// Release the tokens that have already vested.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `token` - Address of the token being released.
+    ///
+    /// # Errors
+    ///
+    /// If the `token` address is not a contract, then the error
+    /// [`Error::InvalidToken`] is returned.
+    /// If the contract fails to execute the call, then the error
+    /// [`Error::ReleaseTokenFailed`] is returned.
+    ///
+    /// # Events
+    ///
+    /// Emits an [`ERC20Released`] event.
+    ///
+    /// # Panics
+    ///
+    /// If `total_allocation * (timestamp - self.start())` exceeds `U256::MAX`.
+    #[selector(name = "release")]
+    fn release_erc20(&mut self, token: Address) -> Result<(), Self::Error>;
+
+    /// Calculates the amount of Ether that has already vested. Default
+    /// implementation is a linear vesting curve.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `timestamp` - Point in time for which to check the vested amount.
+    ///
+    /// # Panics
+    ///
+    /// If total allocation exceeds `U256::MAX`.
+    /// If `total_allocation * (timestamp - self.start())` exceeds `U256::MAX`.
+    #[selector(name = "vestedAmount")]
+    fn vested_amount_eth(&self, timestamp: u64) -> U256;
+
+    /// Calculates the amount of tokens that has already vested. Default
+    /// implementation is a linear vesting curve.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `token` - Address of the token being released.
+    /// * `timestamp` - Point in time for which to check the vested amount.
+    ///
+    /// # Errors
+    ///
+    /// If the `token` address is not a contract, then the error
+    /// [`Error::InvalidToken`] is returned.
+    ///
+    /// # Panics
+    ///
+    /// If `total_allocation * (timestamp - self.start())` exceeds `U256::MAX`.
+    #[selector(name = "vestedAmount")]
+    fn vested_amount_erc20(
+        &mut self,
```

Comment:
This is required to be able to pass it to [Call::new_in](https://github.com/OpenZeppelin/rust-contracts-stylus/pull/402/files/3339d41ac9e4eefefccd4d2d461f370831bd756b#diff-4083d07e30a6c3e16c834ce38c8c2236c57a943e7a89f8e4212209eb4a1bcd96R471). 
Now that I think about it, is reentrancy a problem in this function? :thinking: Maybe using `Call::new()` would be enough here, so that `mut` can be removed.

Reply:
I believe [`Call::new`](https://docs.rs/stylus-sdk/latest/stylus_sdk/call/struct.Call.html#method.new) wdyt @qalisander? Should we also use just `Call::new` in our other contracts?

Reply:
I suspect so. We should try to change